### ISSUE TITLE
HTML5 authenticate_device_async(): HTTPRequest failed!

### ISSUE THREAD

**Issue by JonathanPicques** [heroiclabs/nakama-godot#66]
I'm having trouble to get the HTML5 export to connect to my nakama instance.

When running authenticate_device_async(), nakama-godot returns an "HTTPRequest failed!" exception, but when I look in the network tab, the request was successful (200 OK)

Godot version: 3.3.rc7.official (both desktop exported HTML5 app and https://editor.godotengine.org reports the same bug)

Nakama-godot log:
```
=== Nakama : DEBUG === Request 1 failed with result: 2, response code: 0
```

**Comment by JonathanPicques** [comment]
After further testing, it seems like a godot issue

```python
var http_request := HTTPRequest.new()
add_child(http_request)
var credentials := Marshalls.utf8_to_base64(NAKAMA_SERVER_KEY + ":")
var headers := PoolStringArray()
headers.push_back("Authorization: Basic %s" % credentials)
var body := {"id": "super-device-id"}
http_request.request("https://NAKAMA_HOST/v2/account/authenticate/device?create=true&username=jonathan", headers, true, HTTPClient.METHOD_POST, JSON.print(body))
var result = yield(http_request, "request_completed")
print(result)
```

Prints [2, 0, [], [PoolByteArray]]

**Comment by Faless** [comment]
> Godot version: 3.3.rc7.official

Was it working in 3.2.3? In case, it could be due to changes to the HTML5 fetch API.

**Comment by JonathanPicques** [comment]
It does work indeed on 3.2.3

**Comment by Faless** [comment]
It seems that `fetch` cannot read the `content-encoding` headers from nakama.
I'm unsure why this is the case.

So, if you make a generic fetch on a google, API e.g.:
```javascript
fetch("https://acceleratedmobilepageurl.googleapis.com/$discovery/rest?version=v1")
  .then(r => r.headers.forEach(((v, h) => console.log(h, v))))
```
you'll be able to see the content encoding header:
```
cache-control private
content-encoding gzip
content-length 2433
content-type application/json; charset=UTF-8
date Sat, 03 Apr 2021 09:14:13 GMT
server ESF
vary Origin, X-Origin, Referer
```
i.e.: **`content-encoding gzip`**


To my surprise when making requests to the nakama server I can't seem to read the content-encoding value:
```javascript
fetch("http://192.168.122.147:7350/v2/account/authenticate/device?create=true&")
  .then(r => r.headers.forEach(((v, h) => console.log(h, v))))
```
will output:
```
cache-control no-store, no-cache, must-revalidate
content-length 66
content-type application/json
```
But looking at the request the content encoding header is there:

![no-content-type](https://user-images.githubusercontent.com/1687918/113474541-d2ded900-9470-11eb-8beb-1901c58bf9c6.png)

@novabyte do you have any clue? I'm actually very surprised because the browser does indeed handle the content correctly as gzip, the header seems to just be suppressed in `fetch` (but it's there on other website/webservers)

**Comment by novabyte** [comment]
@Faless We use GRPC and GRPC-Gateway to provide the respective headers in responses to requests made through the gateway. I don't know why this would be a problem but our TypeScript client uses the `window.fetch` API and has no problems with network requests made in the browser.

Maybe @zyro would know whether there's handler logic in the GRPC server which omits the "content-encoding" header?

**Comment by Faless** [comment]
As a follow-up, I think reading headers is limited when doing CORS requests, I was likely trying the google API in a context that didn't trigger the browser CORS restrictions.

I'll investigate further, we might have to scrap `content-length` detection from the godot implementation completely

**Comment by Faless** [comment]
Oh, I see, their API uses [`Access-Control-Expose-Headers`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers) to enable content encoding header in CORS requests.

> I'll investigate further, we might have to scrap `content-length` detection from the godot implementation completely

Sadly, this is going to be the solution in the end (HTML5-only ofc)

**Comment by JonathanPicques** [comment]
Nice catch! Forcing Access-Control-Expose-Headers: * with nginx solves the issue 👍

**Comment by Faless** [comment]
Fixed upstream, fix will be included in upcoming Godot `3.3` release.
